director/director_db.go

```diff
@@ -0,0 +1,133 @@
+/***************************************************************
+ *
+ * Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+package director
+
+import (
+	"embed"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
+
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/server_utils"
+)
+
+type ServerDowntime struct {
+	UUID       string     `gorm:"primaryKey"`
+	Name       string     `gorm:"not null;unique"`
+	FilterType filterType `gorm:"type:text;not null"`
+	// We don't use gorm default gorm.Model to change ID type to string
+	CreatedAt time.Time
+	UpdatedAt time.Time
+}
+
+var db *gorm.DB
+
+//go:embed migrations/*.sql
+var embedMigrations embed.FS
+
+func InitializeDB() error {
+	dbPath := param.Director_DbLocation.GetString()
+	tdb, err := server_utils.InitSQLiteDB(dbPath)
+	if err != nil {
+		return err
+	}
+	db = tdb
+	sqldb, err := db.DB()
+	if err != nil {
+		return errors.Wrapf(err, "Failed to get sql.DB from gorm DB: %s", dbPath)
+	}
+	// Run database migrations
+	if err := server_utils.MigrateDB(sqldb, embedMigrations); err != nil {
+		return err
+	}
+	return nil
+}
+
+func ShutdownDirectorDB() error {
+	return server_utils.ShutdownDB(db)
+}
+
+func CreateServerDowntime(name string, filterType filterType) error {
+	id, err := uuid.NewV7()
+	if err != nil {
+		return errors.Wrap(err, "Unable to create new UUID for new entry in server status table")
+	}
+	serverDowntime := ServerDowntime{
+		UUID:       id.String(),
+		Name:       name,
+		FilterType: filterType,
+		CreatedAt:  time.Now(),
+		UpdatedAt:  time.Now(),
+	}
+
+	if err := db.Create(serverDowntime).Error; err != nil {
+		return err
+	}
+	return nil
+}
+
+func GetServerDowntime(name string) (filterType, error) {
```

Comment:
This comment applies to all of the exported functions in this file other than `InitializeDB`-- think hard about whether we need them to be exported, which implies it's a function that's meant to be called by other packages. In this case, I think the majority of these functions are meant to be used internally by the Director. If you can make the case that some other package may eventually need to call one of these functions(like `launchers` has to do to set up the Director database), keep them exported. Otherwise we should lowercase the first letter of the function.